yax/state/artifact_map.py

```diff
@@ -30,34 +59,82 @@ def __init__(self, artifact_dir, db_fp, exe_graph):
-
         sql = """
             SELECT %s FROM Run WHERE id = ?
         """ % ','.join(params_names)
         with auto_rollback(self.conn) as c:
             c.execute(sql, (run_id,))
-            return dict(zip(input_params, c.fetchone()))
+            values = c.fetchone()
+        if values is None:
+            raise LookupError("Provided `run_id` (%r) does not exist."
+                              % run_id)
+        return dict(zip(input_params, values))
 
     def declare_artifacts(self, config, run_id):
-        for bound_artifact, nodes in self.graph.artifact_dependencies.items():
-            params = self._flatten_config({node.name: config[node.name]
-                                           for node in nodes})
-            rows = self._find_existing_artifacts(bound_artifact, params)
+        """Declare artifacts which are not yet associated with `run_id`.
+
+        If an artifact has already been created with arguments matching its
+        position in the pipeline, associate it with this `run_id`. Otherwise,
+        create a new artifact and associate it with this `run_id`.
+
+        Parameters
+        ----------
+        config : dict of dicts
+            Maps section name to dict of parameters and their arguments.
+        run_id : int
+            The run ID to associate the new (or existing) artifacts with.
+
+        Returns
+        -------
+        None
+
+        """
+        for bound_artifact, nodes in \
+                self.graph.bound_artifact_to_upstream_nodes.items():
+            # Only the params which impact the creation of the artifact need
+            # to be checked for.
+            relevant_params = self._flatten_config(
+                {node.name: config[node.name] for node in nodes})
+
+            rows = self._find_existing_artifacts(bound_artifact,
+                                                 relevant_params)
             if len(rows) == 0:
-                self._declare_new_artifact(run_id, bound_artifact)
+                self._declare_new_artifact(bound_artifact, run_id)
             else:
-                for aid, rid in rows:
-                    if rid == run_id:
+                for a_id, r_id in rows:
+                    if r_id == run_id:
+                        # We've encountered ourselves, something strange must
+                        # have happened during a previous prep
                         continue
                     self._insert_into(
                         'Artifact_Run',
-                        {'run_id': run_id, 'artifact_id': aid})
+                        {'run_id': run_id, 'artifact_id': a_id})
 
-    def _declare_new_artifact(self, run_id, bound_artifact):
+    def _declare_new_artifact(self, bound_artifact, run_id):
+        """Create a fresh artifact, creating the directory and entries."""
         path = os.path.join(self.artifact_dir,
                             "_".join([str(run_id), bound_artifact, 'art']))
         os.mkdir(path)
```

Comment:
Is there some reason you need a handle to 'path'? Why not identify the path and call 'mkdir' at the same time?